gax/src/main/java/com/google/api/gax/batching/DynamicFlowControlSettings.java

```diff
@@ -0,0 +1,155 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.batching;
+
+import com.google.api.core.InternalApi;
+import com.google.api.gax.batching.FlowController.LimitExceededBehavior;
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+import javax.annotation.Nullable;
+
+/** Settings for dynamic flow control */
+@AutoValue
+@InternalApi("For google-cloud-java client use only")
+public abstract class DynamicFlowControlSettings {
+
+  /** Number of outstanding elements that {@link FlowController} allows when it's initiated. */
+  @Nullable
+  public abstract Long getInitialOutstandingElementCount();
+
+  /** Number of outstanding bytes that {@link FlowController} allows when it's initiated. */
+  @Nullable
+  public abstract Long getInitialOutstandingRequestBytes();
+
+  /**
+   * Maximum number of outstanding elements {@link FlowController} allows before enforcing flow
+   * control.
+   */
+  @Nullable
+  public abstract Long getMaxOutstandingElementCount();
+
+  /**
+   * Maximum number of outstanding bytes {@link FlowController} allows before enforcing flow
+   * control.
+   */
+  @Nullable
+  public abstract Long getMaxOutstandingRequestBytes();
+
+  /**
+   * Minimum number of outstanding elements {@link FlowController} allows before enforcing flow
+   * control.
+   */
+  @Nullable
+  public abstract Long getMinOutstandingElementCount();
+
+  /**
+   * Minimum number of outstanding bytes {@link FlowController} allows before enforcing flow
+   * control.
+   */
+  @Nullable
+  public abstract Long getMinOutstandingRequestBytes();
+
+  /** @see FlowControlSettings#getLimitExceededBehavior() */
+  public abstract LimitExceededBehavior getLimitExceededBehavior();
+
+  public abstract Builder toBuilder();
+
+  public static Builder newBuilder() {
+    return new AutoValue_DynamicFlowControlSettings.Builder()
+        .setLimitExceededBehavior(LimitExceededBehavior.Block);
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    public abstract Builder setInitialOutstandingElementCount(Long value);
+
+    public abstract Builder setInitialOutstandingRequestBytes(Long value);
+
+    public abstract Builder setMaxOutstandingElementCount(Long value);
+
+    public abstract Builder setMaxOutstandingRequestBytes(Long value);
+
+    public abstract Builder setMinOutstandingElementCount(Long value);
+
+    public abstract Builder setMinOutstandingRequestBytes(Long value);
+
+    public abstract Builder setLimitExceededBehavior(LimitExceededBehavior value);
+
+    abstract DynamicFlowControlSettings autoBuild();
+
+    public DynamicFlowControlSettings build() {
+      DynamicFlowControlSettings settings = autoBuild();
+      Preconditions.checkArgument(
```

Comment:
Also I think it would be easier to read if you restructured a bit like so:

```java
private void verifyElementCountSettings() {
     bool isEnabled = settings.getInitialOutstandingElementCount() != null || settings.getMinOutstandingElementCount() != null || settings.getMaxOutstandingElementCount() != null;
     if (!isEnabled) { return ; }

      Preconditions.checkNotNull(settings.getMinOutstandingElementCount(), "minOutstandingElementCount must be specified if throttling is enabled");
  Preconditions.checkState(settings.getInitialOutstandingElementCount() <= settings.getMaxOutstandingElementCount(), "initialOustandingElementCount must be less than the max...");
}

private void verifyBytesSettings() {
bool hasBytesThreshold = ...;
 ....
}

```

This would make easier to identify what are all of the requirements